src/gotcha_dl.c

```diff
@@ -5,7 +5,92 @@
 #include "elf_ops.h"
 #include <dlfcn.h>
 
-void* _dl_sym(void* handle, const char* name, void* where);
+/**
+ * structure used to pass lookup addr and return library address.
+ */
+struct Addrs {
+    ElfW(Addr) lookup_addr; // input
+    struct link_map *lmap;  // output
+};
+/**
+ * This is a callback to get headers for each library.
+ * We check if the caller's virtual address is between base address and the virtual addr + memory size.
+ * Then we select the lmap based on base_addr and name
+ *
+ * @param info, information about headers
+ * @param size, size of headers
+ * @param data, data from caller
+ * @return
+ */
+int lib_header_callback(struct dl_phdr_info * info, size_t size, void * data) {
+    struct Addrs* addrs = data;
+    char *name = NULL;
+    ElfW(Addr) load_address;
+    for (int i = 0; i < info->dlpi_phnum; ++i) {
+        if (info->dlpi_phdr[i].p_type == PT_LOAD) {
+            ElfW(Addr) base_addr = info->dlpi_addr + info->dlpi_phdr[i].p_offset;
+            ElfW(Addr) start_addr = base_addr + info->dlpi_phdr[i].p_vaddr;
+            ElfW(Addr) end_addr = start_addr + info->dlpi_phdr[i].p_memsz;
+            if (addrs->lookup_addr >= start_addr && addrs->lookup_addr < end_addr) {
+                name = info->dlpi_name;
+                load_address = info->dlpi_addr;
+                break;
+            }
+        }
+    }
+    if (name) {
+        struct link_map *current = addrs->lmap;
+        while (current) {
+            if (strcmp(current->l_name, name) == 0 && load_address == current->l_addr) {
+                addrs->lmap = current;
+                return 1;
+            }
+            current = current->l_next;
+        }
+    }
+    return 0;
+}
+
+/**
+ * Implement the logic of _dl_sym for supporting RTLD_NEXT
+ * 1. find the caller library using the program headers.
+ * 2. find the second library which has the symbol for RTLD_NEXT
+ * @param handle, handle for dl operation
+ * @param name, name of the symbol
+ * @param who, the virtual address of the caller
+ * @return link_map pointer
+ */
+
+static struct link_map* gotchas_dlsym_rtld_next_lookup(const char *name, void *who) {
+    ElfW(Addr) caller = (ElfW(Addr)) who;
+    /* Iterative over the library headers and find the caller
+     * the address of the caller is set in addrs->library_laddr
+     **/
+    struct Addrs addrs;
+    addrs.lookup_addr = caller;
+    addrs.lmap = _r_debug.r_map;
+    void* symbol;
+    dl_iterate_phdr(lib_header_callback, &addrs);
+    int found_sym = 0;
+    struct link_map *handle = addrs.lmap->l_next;
```

Comment:
So your saying if our addrs.lmap is already NULL then we should throw an error.

Reply:
I'm saying that we need to note when we don't find a match for 'who' and then return an error--whatever the mechanism. You could use addrs.lmap to communicate that. Except as the code is now addrs.lmap is initialized to _r_debug.r_map, so it won't ever be NULL.

(Also, it later occurred to me how tests are passing. The wrong base address calculation bug meant we'd never match a library for 'who'. But this bug means we'll still search for a symbol even when there's no match. So two bugs are canceling each other well enough to pass a test.)

Reply:
```
if (addrs.lmap == NULL) {
        error_printf("RTLD_NEXT used in code not dynamically loaded");
        exit (127);
    }
```

Reply:
When u say "find a match for who". What I understand is that, we have lookup the address of "who" and that could not be matched from my _r_debug. Another interpretation I can think of is, there is some internal structure that we are not able to use which could return NULL based on scope.

Reply:
I'm referring to the first case. We have the lookup address, and no library contains that address.

This could happen if JIT code calls dlsym(RTLD_NEXT, ...).

Reply:
Ok in that case all i have to do is add a int found flag which will be set to one only when the lib is found. Otherwise it will remain 0 as initially set.

Reply:
I will throw an error if caller was not found.

Reply:
fixed.

Reply:
BTW, the address calculation was getting matched before. Even with the incorrect code :). Just that the lookup later was passing as well. I checked this based on debug.

Reply:
Must have been with for segments with '0' offset, which is common but not guaranteed.

Reply:
Yes i checked that was the case for me :).